### PR TITLE
Cmmd fuse2

### PR DESCRIPTION
**Pull request by itaijj**

### CODE REVIEW

File: examples/fuse_examples/tests/test_classification_cmmd.py [hunk 1]
```diff
@@ -34,53 +35,40 @@
...
         self.assertGreaterEqual(results['metrics.auc'], threshold)
 
-    def tearDown(self):
-        # Delete temporary directories
-        shutil.rmtree(self.ROOT)
-        shutil.rmtree(self.CACHE_PATH)
-
-
-if __name__ == '__main__':
+    # def tearDown(self):
```
Comment (mosheraboh) [inline examples/fuse_examples/tests/test_classification_cmmd.py:63]: Why?

File: fuseimg/data/ops/shape_ops.py [hunk 2]
```diff
@@ -82,7 +81,214 @@ def __call__(self, sample_dict: NDict, op_id: Optional[str], key: str,
...
+                                                   anti_aliasing=True
+                                                   )
+        return inner_image
+    
+    def __call__(self, sample_dict: NDict, key: str , resize_to: Tuple,
+                 padding: Tuple
+        ):
+        '''
+        :param resize_to:               new size of input images, keeping proportions
+        :param padding:                 padding size
```
Comment (mosheraboh) [inline fuseimg/data/ops/shape_ops.py:170]: What is the format of padding -  there are few possible options.

File: fuseimg/data/ops/shape_ops.py [hunk 3]
```diff
@@ -82,7 +81,214 @@ def __call__(self, sample_dict: NDict, op_id: Optional[str], key: str,
...
+class OpRemoveDarkBackgroundRectangle2D(OpBase):
+    '''
+     removes dark background in a rectangle patch from a 2d image
+    '''
+    def __init__(self, dark_area_threshold : int = 10, blocks_num : int = 30, **kwargs):
+        super().__init__(**kwargs)
+        self.dark_area_threshold = dark_area_threshold
+        self.blocks_num = blocks_num
+
+    def find_breast_aabb(self,img : np.ndarray) -> Tuple[int ,int, int, int]:
```
Comment (mosheraboh) [inline fuseimg/data/ops/shape_ops.py:193]: is it relevant just for breast?

File: fuseimg/data/ops/shape_ops.py [hunk 4]
```diff
@@ -82,7 +81,214 @@ def __call__(self, sample_dict: NDict, op_id: Optional[str], key: str,
...
+        cells = np.zeros((self.blocks_num,self.blocks_num))
+
+        for ri,r in enumerate(rows_starts):
+            for ci,c in enumerate(cols_starts):
+                r_end = min(img.shape[0]-1,r+bl_rows)
+                c_end = min(img.shape[1] - 1, c + bl_cols)
+                cells[ri,ci] = np.mean(img[r:r_end,c:c_end])
+
+        cells_binary = np.zeros((self.blocks_num,self.blocks_num),dtype=bool)
+        cells_binary[cells>self.dark_area_threshold] = True #TODO: this is a hardcoded threshold, see if there's a better alternative
```
Comment (mosheraboh) [inline fuseimg/data/ops/shape_ops.py:221]: missed TODO here

File: fuseimg/data/ops/shape_ops.py [hunk 5]
```diff
@@ -82,7 +81,214 @@ def __call__(self, sample_dict: NDict, op_id: Optional[str], key: str,
...
+        if resize_to is not None:
+            img = self.resize_image(img, resize_to)
+
+        # padding
+        if padding is not None:
+            img = self.pad_image(img, padding, resize_to)
+        sample_dict[key] = img
+        return sample_dict
+        
+class OpRemoveDarkBackgroundRectangle2D(OpBase):
```
Comment (mosheraboh) [inline fuseimg/data/ops/shape_ops.py:184]: Maybe you can explain do you mean by remove?
It crops the image, trying to remove the background in the outline?

File: fuseimg/data/ops/shape_ops.py [hunk 6]
```diff
@@ -82,7 +81,214 @@ def __call__(self, sample_dict: NDict, op_id: Optional[str], key: str,
...
         
+        img = sample_dict[key]
+        aabb = self.find_breast_aabb(img)
+        img = img[aabb[0]: aabb[2], aabb[1]: aabb[3]].copy()
+        sample_dict[key] = img
+        return sample_dict
+    
+class OpFlipBrightSideOnLeft2D(OpBase):
+    '''
+     checks if the image needed to be flipped to the left
```
Comment (mosheraboh) [inline fuseimg/data/ops/shape_ops.py:259]: How do you check? is it relevant just for MGs? explain here

File: fuseimg/data/ops/shape_ops.py [hunk 7]
```diff
@@ -82,7 +81,214 @@ def __call__(self, sample_dict: NDict, op_id: Optional[str], key: str,
...
+    '''
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.max_pixel_value = 255.0
+        self.dark_region_ratio = 15.0
```
Comment (mosheraboh) [inline fuseimg/data/ops/shape_ops.py:264]: Why is it hardcoded?

File: fuseimg/data/ops/shape_ops.py [hunk 8]
```diff
@@ -82,7 +81,214 @@ def __call__(self, sample_dict: NDict, op_id: Optional[str], key: str,
...
+        sample_dict[key] = img
+        return sample_dict
+    
+class OpFlipBrightSideOnLeft2D(OpBase):
+    '''
+     checks if the image needed to be flipped to the left
+    '''
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.max_pixel_value = 255.0
```
Comment (mosheraboh) [inline fuseimg/data/ops/shape_ops.py:263]: Do you expect the image to be between 0:255? Do expect a gray scale?.
Specify all your assumptions in the comments please.

File: fuseimg/data/ops/shape_ops.py [hunk 9]
```diff
@@ -82,7 +81,214 @@ def __call__(self, sample_dict: NDict, op_id: Optional[str], key: str,
...
+        :param padding: required padding [x,y]
+        :param resize_to: original requested resolution
+        :param normalized_target_range: requested normalized image pixels range
+        :param number_of_channels: number of color channels in the image
+        :return: padded image
+        """
+        inner_image = inner_image.astype('float32')
+        # "Pad" around inner image
+        inner_image_height, inner_image_width = inner_image.shape[0], inner_image.shape[1]
+        inner_image[0:inner_image_height, 0] = 0
```
Comment (mosheraboh) [inline fuseimg/data/ops/shape_ops.py:108]: Can you explain?

File: fuseimg/data/ops/shape_ops.py [hunk 10]
```diff
@@ -82,7 +81,214 @@ def __call__(self, sample_dict: NDict, op_id: Optional[str], key: str,
...
+        pads both side equally by the same input padding size (left = right = padding[1] , up = down= padding[0] )  ,
+        padding default value is zero or minimum value in normalized target range
+        :param inner_image: image of shape [H, W, C]
+        :param padding: required padding [x,y]
+        :param resize_to: original requested resolution
+        :param normalized_target_range: requested normalized image pixels range
+        :param number_of_channels: number of color channels in the image
+        :return: padded image
+        """
+        inner_image = inner_image.astype('float32')
```
Comment (mosheraboh) [inline fuseimg/data/ops/shape_ops.py:105]: Don't force it. Specify that you expect float32 image if it necessary

File: fuseimg/data/ops/image_loader.py [hunk 11]
```diff
@@ -35,3 +35,33 @@ def __call__(self, sample_dict: NDict, op_id: Optional[str], key_in:str, key_out
...
+
+    def __call__(self, sample_dict: NDict, key_in:str, key_out: str, format:str="infer"):
+        '''
+        :param key_in: the key name in sample_dict that holds the filename
+        :param key_out: 
+        '''
+        img_filename = os.path.join(self._dir_path, sample_dict[key_in])
+        img_filename_suffix = img_filename.split(".")[-1]
+        if (format == "infer" and img_filename_suffix in ["dcm"]) or \
+            (format in ["nii", "nib"]):
```
Comment (mosheraboh) [inline fuseimg/data/ops/image_loader.py:54]: format in "dcm"

File: fuseimg/data/ops/image_loader.py [hunk 12]
```diff
@@ -35,3 +35,33 @@ def __call__(self, sample_dict: NDict, op_id: Optional[str], key_in:str, key_out
     def reverse(self, sample_dict: dict, key_to_reverse: str, key_to_follow: str, op_id: Optional[str]) -> dict:
         return sample_dict
 
+class OpLoadDicom(OpBase):
+    '''
+    Loads a medical image, currently only nii is supported
```
Comment (mosheraboh) [inline fuseimg/data/ops/image_loader.py:40]: only dcm format

File: fuseimg/data/ops/image_loader.py [hunk 13]
```diff
@@ -35,3 +35,33 @@ def __call__(self, sample_dict: NDict, op_id: Optional[str], key_in:str, key_out
...
+    '''
+    def __init__(self, dir_path: str , **kwargs):
+        super().__init__(**kwargs)
+        self._dir_path = dir_path
+
+    def __call__(self, sample_dict: NDict, key_in:str, key_out: str, format:str="infer"):
+        '''
+        :param key_in: the key name in sample_dict that holds the filename
+        :param key_out:
```
Comment (mosheraboh) [inline fuseimg/data/ops/image_loader.py:49]: Do you extract just the image?
Do support an option to read more fields from the dicom? (it will be useful)

File: fuseimg/data/ops/image_loader.py [hunk 14]
```diff
@@ -35,3 +35,33 @@ def __call__(self, sample_dict: NDict, op_id: Optional[str], key_in:str, key_out
     def reverse(self, sample_dict: dict, key_to_reverse: str, key_to_follow: str, op_id: Optional[str]) -> dict:
         return sample_dict
 
+class OpLoadDicom(OpBase):
+    '''
+    Loads a medical image, currently only nii is supported
+    '''
+    def __init__(self, dir_path: str , **kwargs):
```
Comment (mosheraboh) [inline fuseimg/data/ops/image_loader.py:42]: document arguments

**Review by mosheraboh — CHANGES_REQUESTED** [review]
Thanks Itai.
Please find my comments inline

File: examples/fuse_examples/imaging/classification/cmmd/conf/config.yaml [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+train:
+  force_reset_model_dir: True
+  reset_cache: False
+  num_workers : 8
+  batch_size: 14
+  learning_rate : 1e-5
+  weight_decay :  0.001
+  resume_checkpoint_filename : 
+  manager_train_params:
+    num_gpus : 7
```
[the remark this answers is not in the document]
Comment (itaijj) [reply examples/fuse_examples/imaging/classification/cmmd/conf/config.yaml:17]: it's just supposed to be an just an example file not a recommendation

File: examples/fuse_examples/imaging/classification/cmmd/dataset.py [hunk 16]
```diff
@@ -1,106 +1,168 @@
...
+from fuse.data.ops.ops_aug_common import OpSample
+from fuse.data.ops.ops_read import OpReadDataframe
+import torch
+from fuse.utils.rand.param_sampler import RandBool, RandInt, Uniform
 from fuse.utils.rand.param_sampler import Uniform, RandInt, RandBool
+import numpy as np
+from fuse.data.utils.split import SplitDataset
+from typing import Tuple, List
+
+def create_folds(input_source: str,
```
[the remark this answers is not in the document]
Comment (itaijj) [reply examples/fuse_examples/imaging/classification/cmmd/dataset.py:27]: ok, thanks

File: examples/fuse_examples/imaging/classification/cmmd/dataset.py [hunk 17]
```diff
@@ -1,106 +1,168 @@
...
+                                                                    nfolds = num_folds ,
+                                                                    print_flag=True )
+                # Extract entities
+                # ----------------
+            else:
+                create_folds.folds_df = pd.read_csv(partition_file_name)
+
+    return create_folds.folds_df[create_folds.folds_df['fold'].isin(folds)]
+
+def create_dataset_partition(phase : str,
```
[the remark this answers is not in the document]
Comment (itaijj) [reply examples/fuse_examples/imaging/classification/cmmd/dataset.py]: I moved it but didn't update git somehow, 
I don't think we can have exact similar interface

File: examples/fuse_examples/tests/test_classification_cmmd.py (hunk 1 above)
Comment (itaijj) [reply examples/fuse_examples/tests/test_classification_cmmd.py:63]: I couldn't make it run still, have some issues with running folder

File: fuseimg/data/ops/shape_ops.py (hunk 2 above)
Comment (itaijj) [reply fuseimg/data/ops/shape_ops.py:170]: do you mean how much you pad in each size? 
right now we have 2 values for each axis, and it is symmetric on each size

Comment (itaijj) [reply fuseimg/data/ops/shape_ops.py:170]: it's here in the documentation also
        pads image to requested size ,
        pads both side equally by the same input padding size (left = right = padding[1] , up = down= padding[0] )  ,
        padding default value is zero or minimum value in normalized target range

File: fuseimg/data/ops/shape_ops.py (hunk 3 above)
Comment (itaijj) [reply fuseimg/data/ops/shape_ops.py:193]: will fix the name

File: fuseimg/data/ops/shape_ops.py (hunk 4 above)
Comment (itaijj) [reply fuseimg/data/ops/shape_ops.py:221]: it's something we inherit from old code, you want to just remove it ?

File: fuseimg/data/ops/shape_ops.py (hunk 5 above)
Comment (itaijj) [reply fuseimg/data/ops/shape_ops.py]: I renamed it, it finds the biggest non empty bbox

File: fuseimg/data/ops/shape_ops.py (hunk 6 above)
Comment (itaijj) [reply fuseimg/data/ops/shape_ops.py:258]: removed all reference to breast and MG

File: fuseimg/datasets/cmmd.py [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+        Get suggested static pipeline (which will be cached), typically loading the data plus design choices that we won't experiment with.
+        :param data_path: path to original kits21 data (can be downloaded by KITS21.download())
+        """
+        static_pipeline = PipelineDefault("static", [
+            
+            (OpReadDataframe(data_source,key_column = None , columns_to_extract = ['file','classification'] , rename_columns=dict(file="data.input.img_path",classification="data.gt.classification")), dict()), # will save image and seg path to "data.input.img_path", "data.gt.seg_path" 
+            (OpLoadDicom(data_dir), dict(key_in="data.input.img_path", key_out="data.input.img", format="nib")),
+            (OpFlipBrightSideOnLeft2D(), dict(key="data.input.img")),
+            (OpFindBiggestNonEmptyBbox2D(), dict(key="data.input.img")),
+            (OpNormalizeAgainstSelfImpl(), dict(key="data.input.img")),
```
Comment (mosheraboh) [inline fuseimg/datasets/cmmd.py:47]: I know that it was already like this.
But can you please remove the "impl" suffix?

File: fuseimg/datasets/cmmd.py [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+                    )),
+            (OpLambda(partial(torch.unsqueeze, dim=0)), dict(key="data.input.img")), 
+            
+        ])
+        return dynamic_pipeline
+
+    @staticmethod
+    def create_dataset_partition(phase : str,
+                                data_dir: str,
+                                data_source : pd.DataFrame,
```
Comment (mosheraboh) [inline fuseimg/datasets/cmmd.py:81]: Why do you need data_source and specific_ids? it's the same, right?
Can you rename the function name to dataset?

Comment (itaijj) [reply fuseimg/datasets/cmmd.py:81]: I used for debug, because of this one problematic image.
I can remove it and only use data_source

File: fuseimg/datasets/cmmd.py (hunk 18 above)
Comment (itaijj) [reply fuseimg/datasets/cmmd.py:47]: no problem

File: examples/fuse_examples/imaging/classification/cmmd/conf/config.yaml (hunk 15 above)
[the remark this answers is not in the document]
Comment (mosheraboh) [reply examples/fuse_examples/imaging/classification/cmmd/conf/config.yaml]: Shouldn't be the baseline hyperparameters?
Anyway, it's better to have something that someone can run.
I would set it to 1 GPU unless you can get better results with 2 GPUs

File: examples/fuse_examples/imaging/classification/cmmd/dataset.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (mosheraboh) [reply examples/fuse_examples/imaging/classification/cmmd/dataset.py]: Thanks

File: fuseimg/data/ops/shape_ops.py (hunk 7 above)
Comment (itaijj) [reply fuseimg/data/ops/shape_ops.py:263]: changes to parameter

File: fuseimg/data/ops/shape_ops.py (hunk 8 above)
Comment (itaijj) [reply fuseimg/data/ops/shape_ops.py:262]: grayscale, it was already in the parameter documentation

File: fuseimg/data/ops/image_loader.py (hunk 11 above)
Comment (itaijj) [reply fuseimg/data/ops/image_loader.py:54]: ok

File: fuseimg/data/ops/image_loader.py (hunk 12 above)
Comment (itaijj) [reply fuseimg/data/ops/image_loader.py:40]: ok

File: fuseimg/data/ops/shape_ops.py (hunk 9 above)
Comment (itaijj) [reply fuseimg/data/ops/shape_ops.py:108]: added more documentation

File: fuseimg/data/ops/shape_ops.py (hunk 10 above)
Comment (itaijj) [reply fuseimg/data/ops/shape_ops.py:105]: removed

File: fuseimg/data/ops/image_loader.py (hunk 14 above)
Comment (itaijj) [reply fuseimg/data/ops/image_loader.py:42]: ok

File: fuseimg/data/ops/image_loader.py (hunk 13 above)
Comment (itaijj) [reply fuseimg/data/ops/image_loader.py:49]: I added basic field reading

File: fuseimg/data/ops/image_loader.py [hunk 20]
```diff
@@ -37,29 +37,39 @@ def reverse(self, sample_dict: dict, key_to_reverse: str, key_to_follow: str, op
...
     '''
     def __init__(self, dir_path: str , **kwargs):
+        '''
+        :param dir_path: the main folder path where the images located
+        '''
         super().__init__(**kwargs)
         self._dir_path = dir_path
 
-    def __call__(self, sample_dict: NDict, key_in:str, key_out: str, format:str="infer"):
+    def __call__(self, sample_dict: NDict, key_in:str, key_out: str, format:str="infer", metadata_keys: NDict=None, key_metadata_out: str="data.input.metadata"):
```
Comment (mosheraboh) [inline fuseimg/data/ops/image_loader.py:49]: add documentation of the new args
key_metadata_out default should be None, for a case that you don't want to read argments

Comment (mosheraboh) [reply fuseimg/data/ops/image_loader.py:49]: metadata_keys could be a simple dict (no need for NDict here)

File: fuseimg/data/ops/shape_ops.py [hunk 21]
```diff
@@ -95,20 +95,23 @@ def pad_image(self,inner_image: np.ndarray, padding: Tuple[float, float], resize
...
         """
-        inner_image = inner_image.astype('float32')
-        # "Pad" around inner image
+        # "Pad" around inner image with background value 
         inner_image_height, inner_image_width = inner_image.shape[0], inner_image.shape[1]
-        inner_image[0:inner_image_height, 0] = 0
-        inner_image[0:inner_image_height, inner_image_width - 1] = 0
-        inner_image[0, 0:inner_image_width] = 0
-        inner_image[inner_image_height - 1, 0:inner_image_width] = 0
+        # pad first row
```
Comment (mosheraboh) [inline fuseimg/data/ops/shape_ops.py:107]: Why do you need to pad here if you apply padding 116, is it because you are using reflect method?

File: fuseimg/datasets/cmmd.py [hunk 22]
```diff
@@ -76,20 +79,18 @@ def dynamic_pipeline():
...
+    def dataset(phase : str,
+                data_dir: str,
+                data_source : pd.DataFrame,
+                cache_dir : str = None,
+                restart_cache : bool = True) :
         """
         Creates Fuse Dataset single object (either for training, validation and test or user defined set)
         :param phase:                       parition name (training / validation / test)
         :param data_dir:                    dataset root path
         :param data_source                  csv file containing all samples file paths and ground truth
```
Comment (mosheraboh) [inline fuseimg/datasets/cmmd.py:91]: Can you get sample_ids and not csv file?

**Review by mosheraboh — CHANGES_REQUESTED** [review]

File: fuseimg/data/ops/image_loader.py (hunk 20 above)
Comment (itaijj) [reply fuseimg/data/ops/image_loader.py:49]: fixed

File: fuseimg/datasets/cmmd.py (hunk 22 above)
Comment (itaijj) [reply fuseimg/datasets/cmmd.py:91]: I won't know how to locate the image like this

File: examples/fuse_examples/imaging/classification/cmmd/runner.py [hunk 23]
```diff
@@ -43,11 +45,12 @@
...
 from fuse.dl.managers.manager_default import ManagerDefault
-
-from fuse_examples.imaging.classification.cmmd.dataset import CMMD_2021_dataset
+from fuseimg.datasets.cmmd import CMMD
 from fuse.dl.models.backbones.backbone_inception_resnet_v2 import BackboneInceptionResnetV2
 
-
+import hydra
+from typing import Dict
+from omegaconf import DictConfig, OmegaConf
```
Comment (mosheraboh) [inline examples/fuse_examples/imaging/classification/cmmd/runner.py:53]: Is it part of hydra package? should we add it to requirements as well

File: examples/fuse_examples/tests/test_classification_cmmd.py [hunk 24]
```diff
@@ -34,53 +35,40 @@
...
-        EXPERIMENT_CACHE = 'CMMD_cache'
-
-        self.paths = {'data_dir': self.ROOT_DATA,
-                      'model_dir': os.path.join(self.ROOT, EXPERIMENT, 'model_dir'),
-                      'data_misc_dir' : os.path.join(self.ROOT, 'data_misc'),
-                      'force_reset_model_dir': True,
-                      'cache_dir': os.path.join(self.CACHE_PATH, EXPERIMENT_CACHE + '_cache_dir'),
-                      'inference_dir': os.path.join(self.ROOT, EXPERIMENT, 'infer_dir'),
-                      'eval_dir': os.path.join(self.ROOT, EXPERIMENT, 'eval_dir')}
+        global config
```
[the remark this answers is not in the document]
Comment (mosheraboh) [reply examples/fuse_examples/tests/test_classification_cmmd.py:39]: Ohh - you probably working on it now, we'll skip for now.

File: fuseimg/data/ops/image_loader.py [hunk 25]
```diff
@@ -27,18 +25,53 @@ def __call__(self, sample_dict: NDict, op_id: Optional[str], key_in:str, key_out
             (format in ["nii", "nib"]):  
             img = nib.load(img_filename)
             img_np = img.get_fdata()
-            sample_dict[key_out] = img_np
```
Comment (mosheraboh) [inline fuseimg/data/ops/image_loader.py:30]: Why did you remove it?

**Review by mosheraboh — COMMENTED** [review]
Almost done 👍

File: examples/fuse_examples/imaging/classification/cmmd/conf/config.yaml [hunk 26]
```diff
@@ -0,0 +1,37 @@
+paths:
+  data_dir: /projects/msieve3/CMMD
```
[the remark this answers is not in the document]
Comment (itaijj) [reply examples/fuse_examples/imaging/classification/cmmd/conf/config.yaml:2]: i'm using a different hydra file there so it's not a problem.
not sure what you mean by environment variable

File: examples/fuse_examples/imaging/classification/cmmd/runner.py [hunk 27]
```diff
@@ -56,56 +59,14 @@
...
-    # can be either better/worse - whether to consider best epoch when values are equal
-}
-TRAIN_COMMON_PARAMS['manager.learning_rate'] = 1e-5
-TRAIN_COMMON_PARAMS['manager.weight_decay'] = 0.001
-TRAIN_COMMON_PARAMS['manager.resume_checkpoint_filename'] = None
 #################################
 # Train Template
 #################################
-def run_train(paths: dict, train_common_params: dict, reset_cache: bool):
+def run_train(paths : NDict , train: NDict ):
```
[the remark this answers is not in the document]
Comment (itaijj) [reply examples/fuse_examples/imaging/classification/cmmd/runner.py:65]: it's just easier later to use it as in hydra X.Y.Z

File: examples/fuse_examples/imaging/classification/cmmd/runner.py [hunk 28]
```diff
@@ -117,56 +78,84 @@ def run_train(paths: dict, train_common_params: dict, reset_cache: bool):
...
+
+    train_sample_ids = []
+    for fold in train["train_folds"]:
+        train_sample_ids += folds[fold]
+    validation_sample_ids = []
+    for fold in train["validation_folds"]:
+        validation_sample_ids += folds[fold]
+
+    train_dataset = CMMD.dataset(paths["data_dir"], paths["data_misc_dir"], train['target'], paths["cache_dir"], reset_cache=False, num_workers=train["num_workers"], sample_ids=train_sample_ids, train=True)
+    # for _ in train_dataset:
```
[the remark this answers is not in the document]
Comment (itaijj) [reply examples/fuse_examples/imaging/classification/cmmd/runner.py:110]: done

File: examples/fuse_examples/imaging/classification/cmmd/runner.py [hunk 29]
```diff
@@ -298,57 +290,40 @@ def run_eval(paths: dict, eval_common_params: dict):
...
     force_gpus = None  # [0]
-    choose_and_enable_multiple_gpus(NUM_GPUS, force_gpus=force_gpus)
+    choose_and_enable_multiple_gpus(cfg["train.manager_train_params.num_gpus"], force_gpus=force_gpus)
 
-    RUNNING_MODES = ['train', 'infer', 'eval']  # Options: 'train', 'infer', 'eval'
-    # Path to save model
-    root = ''
-    # Path to the stored CMMD dataset location
+
+    RUNNING_MODES = ['train', 'infer', 'analyze']  # Options: 'train', 'infer', 'analyze'
```
[the remark this answers is not in the document]
Comment (itaijj) [reply examples/fuse_examples/imaging/classification/cmmd/runner.py:308]: done

File: fuse/data/utils/export.py [hunk 30]
```diff
@@ -59,6 +59,9 @@ def export_to_dataframe(dataset: DatasetBase, keys: Sequence[str], output_filena
             values = [sample_dict[key] for sample_dict in data]
             df[key] = values
 
+        # set sample_id as index
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuse/data/utils/export.py:62]: we did it together, it's because here we need the original ID ( file name) to be a column and not index

File: fuseimg/data/ops/image_loader.py [hunk 31]
```diff
@@ -27,18 +25,53 @@ def __call__(self, sample_dict: NDict, op_id: Optional[str], key_in:str, key_out
...
+    Loads a medical image, currently only dicom is supported
+    '''
+    def __init__(self, dir_path: str , **kwargs):
+        '''
+        :param dir_path: the main folder path where the images located
+        '''
+        super().__init__(**kwargs)
+        self._dir_path = dir_path
+
+    def __call__(self, sample_dict: NDict, key_in:str, key_out: str, format: str="infer", metadata_keys: Dict=None, key_metadata_out: str=None):
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuseimg/data/ops/image_loader.py:49]: ok

File: fuseimg/data/ops/image_loader.py (hunk 25 above)
Comment (itaijj) [reply fuseimg/data/ops/image_loader.py:30]: it's an old code I had

File: fuseimg/data/ops/image_loader.py (hunk 31 above)
[the remark this answers is not in the document]
Comment (itaijj) [reply fuseimg/data/ops/image_loader.py:49]: done

File: fuseimg/data/ops/image_loader.py [hunk 32]
```diff
@@ -27,18 +25,53 @@ def __call__(self, sample_dict: NDict, op_id: Optional[str], key_in:str, key_out
...
+        '''
+        :param key_in: the key name in sample_dict that holds the filename
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuseimg/data/ops/image_loader.py:51]: merged this op with opLoadImage

File: examples/fuse_examples/imaging/classification/cmmd/runner.py [hunk 33]
```diff
@@ -298,57 +290,40 @@ def run_eval(paths: dict, eval_common_params: dict):
...
+    # Path to the stored dataset location
     # dataset should be download from https://wiki.cancerimagingarchive.net/pages/viewpage.action?pageId=70230508
     # download requires NBIA data retriever https://wiki.cancerimagingarchive.net/display/NBIA/Downloading+TCIA+Images
-    # put on the following in the main folder  - 
+    # put on the folliwing in the main folder  -
```
[the remark this answers is not in the document]
Comment (itaijj) [reply examples/fuse_examples/imaging/classification/cmmd/runner.py:312]: fixed

File: fuse/data/utils/export.py (hunk 30 above)
[the remark this answers is not in the document]
Comment (itaijj) [reply fuse/data/utils/export.py:62]: we did it together because of the fold partition in my case where index!= id

**Review by mosheraboh — COMMENTED** [review]
@YoelShoshan , can you please review just the hydra part in this example?

File: examples/fuse_examples/tests/test_classification_cmmd.py [hunk 34]
```diff
@@ -17,70 +17,72 @@
...
+        # Path to the stored dataset location
+        # dataset should be download from https://wiki.cancerimagingarchive.net/pages/viewpage.action?pageId=70230508
+        # download requires NBIA data retriever https://wiki.cancerimagingarchive.net/display/NBIA/Downloading+TCIA+Images
+        # put on the folliwing in the main folder  - 
+        # 1. CMMD_clinicaldata_revision.csv which is a converted version of CMMD_clinicaldata_revision.xlsx 
+        # 2. folder named CMMD which is the downloaded data folder
 
-        self.eval_common_params = EVAL_COMMON_PARAMS
 
     def test_runner(self):
```
Comment (mosheraboh) [inline examples/fuse_examples/tests/test_classification_cmmd.py:68]: You should add a decorator here:
`@run_in_subprocess()`

Comment (itaijj) [reply examples/fuse_examples/tests/test_classification_cmmd.py:68]: done

File: examples/fuse_examples/tests/test_classification_cmmd.py [hunk 35]
```diff
@@ -74,14 +76,14 @@ def test_runner(self):
...
-        threshold = 0.6
-        self.assertGreaterEqual(results['metrics.auc'], threshold)
+        self.assertTrue('metrics.auc' in results)
 
     def tearDown(self):
         # Delete temporary directories
         shutil.rmtree(self.root)
         
 def main() -> None:
+    # multiprocessing.set_start_method('spawn')
```
Comment (mosheraboh) [inline examples/fuse_examples/tests/test_classification_cmmd.py:86]: remove

File: fuse/data/utils/export.py [hunk 36]
```diff
@@ -60,7 +60,7 @@ def export_to_dataframe(dataset: DatasetBase, keys: Sequence[str], output_filena
             df[key] = values
 
         # set sample_id as index
-        df = df.set_index(sample_id_key, drop=False)
+        # df = df.set_index(sample_id_key, drop=False)
```
Comment (mosheraboh) [inline fuse/data/utils/export.py:63]: remove

File: examples/fuse_examples/tests/test_classification_cmmd.py (hunk 35 above)
Comment (itaijj) [reply examples/fuse_examples/tests/test_classification_cmmd.py:86]: done

File: fuse/data/utils/export.py (hunk 36 above)
Comment (itaijj) [reply fuse/data/utils/export.py:63]: done

File: examples/fuse_examples/tests/test_classification_cmmd.py [hunk 37]
```diff
@@ -17,70 +17,74 @@
...
-
-
+import tempfile
+import os
 from fuse.utils.gpu import choose_and_enable_multiple_gpus
+from fuse.utils.multiprocessing.run_multiprocessed import run_in_subprocess
+assert "CMMD_DATA_PATH" in os.environ, "Expecting environment variable CMMD_DATA_PATH to be set. Follow the instruction in example README file to download and set the path to the data"
 
-@unittest.skip("Not ready yet")
+# @unittest.skip("Not ready yet")
```
Comment (mosheraboh) [inline examples/fuse_examples/tests/test_classification_cmmd.py:31]: remove, keep the skip as in stoic21

File: examples/fuse_examples/tests/test_classification_cmmd.py [hunk 38]
```diff
@@ -17,70 +17,74 @@
...
-import tempfile
 import shutil
-
-
-
+import tempfile
+import os
 from fuse.utils.gpu import choose_and_enable_multiple_gpus
+from fuse.utils.multiprocessing.run_multiprocessed import run_in_subprocess
+assert "CMMD_DATA_PATH" in os.environ, "Expecting environment variable CMMD_DATA_PATH to be set. Follow the instruction in example README file to download and set the path to the data"
```
Comment (mosheraboh) [inline examples/fuse_examples/tests/test_classification_cmmd.py:29]: remove

File: fuse/data/utils/split.py [hunk 39]
```diff
@@ -139,7 +139,7 @@ def dataset_balanced_division_to_folds(dataset: DatasetBase, output_split_filena
         if keys_to_balance is not None:
             keys += keys_to_balance
         df = ExportDataset.export_to_dataframe(dataset, keys, workers = workers, mp_context=mp_context,  **kwargs)
-        df_folds = balanced_division(df, id, keys_to_balance, nfolds, **kwargs)
+        df_folds = balanced_division(df, id, keys_to_balance, nfolds)
```
Comment (mosheraboh) [inline fuse/data/utils/split.py:142]: You should add back kwargs here and remove it in line 141

**Review by mosheraboh — APPROVED** [review]
Thanks Itai!
I'm merging, please address the last comment in new PR.

File: examples/fuse_examples/imaging/classification/cmmd/runner.py [hunk 40]
```diff
@@ -298,57 +294,41 @@ def run_eval(paths: dict, eval_common_params: dict):
...
     force_gpus = None  # [0]
-    choose_and_enable_multiple_gpus(NUM_GPUS, force_gpus=force_gpus)
+    choose_and_enable_multiple_gpus(cfg["train.manager_train_params.num_gpus"], force_gpus=force_gpus)
+
 
-    RUNNING_MODES = ['train', 'infer', 'eval']  # Options: 'train', 'infer', 'eval'
-    # Path to save model
-    root = ''
-    # Path to the stored CMMD dataset location
+    RUNNING_MODES = ['train', 'infer', 'eval']
```
Comment (YoelShoshan) [inline examples/fuse_examples/imaging/classification/cmmd/runner.py:313]: maybe this should be part of the hydra config as well?

Comment (itaijj) [reply examples/fuse_examples/imaging/classification/cmmd/runner.py:313]: accepted
